shesmu-server-ui/src/html.ts

```diff
@@ -0,0 +1,1618 @@
+ */
+export type UIElement = UIElement[] | string | Node;
+
+/**
+ * A list that can be updated
+ */
+export interface UpdateableList<T> {
+  add(item: T): void;
+  keepOnly(predicate: (item: T) => boolean): void;
+  replace(items: T[]): void;
+}
+let activeMenu: HTMLElement | null = null;
+let closeActiveMenu: (isExternal: boolean) => void = (v) => {};
+let findOverride: FindHandler = null;
+
+/**
+ * Add all GUI elements to an existing HTML element
+ */
+export function addElements(
+  target: HTMLElement,
+  ...elements: UIElement[]
+): void {
+  elements.flat(Number.MAX_VALUE).forEach((result: string | Node) => {
+    if (typeof result == "string") {
+      target.appendChild(document.createTextNode(result));
+    } else {
+      target.appendChild(result);
+    }
+  });
+}
+
+/**
+ * Add a blank UI element
+ *
+ * This isn't very useful, but simplifies some code paths in the action tile building.
+ */
+export function blank(): UIElement {
+  return [];
+}
+/**
+ * Create a line break
+ */
+export function br(): UIElement {
+  return document.createElement("br");
+}
+
+/**
+ * Create a modal dialog with a throbber in the middle.
+ *
+ * It returns a function to close this dialog
+ */
+export function busyDialog(): () => void {
+  const modal = document.createElement("div");
+  modal.className = "modal";
+  addElements(modal, throbber());
+  document.body.appendChild(modal);
+  return () => document.body.removeChild(modal);
+}
+
+/**
+ * Create a normal button
+ */
+export function button(
+  label: UIElement,
+  title: string,
+  callback: ClickHandler
+) {
+  return buttonCustom(label, title, [], callback);
+}
+
+/**
+ * Create a button for an unimportant feature
+ */
+export function buttonAccessory(
+  label: UIElement,
+  title: string,
+  callback: ClickHandler
+) {
+  return buttonCustom(label, title, ["accessory"], callback);
+}
+
+/**
+ * Create a button for with a custom UI design
+ */
+export function buttonCustom(
+  label: UIElement,
+  title: string,
+  className: string[],
+  callback: ClickHandler
+) {
+  const button = document.createElement("span");
+  button.classList.add("load");
+  for (const name of className) {
+    button.classList.add(name);
+  }
+  addElements(button, label);
+  button.title = title;
+  button.addEventListener("click", callback);
+  return button;
+}
+
+/**
+ * Create a close button
+ */
+export function buttonClose(title: string, callback: ClickHandler): UIElement {
+  return buttonSubtle("✖", title, callback);
+}
+/**
+ * Create a button for a dangerous feature
+ */
+export function buttonDanger(
+  label: UIElement,
+  title: string,
+  callback: ClickHandler
+) {
+  return buttonCustom(label, title, ["danger"], callback);
+}
+/**
+ * Create an edit button
+ */
+export function buttonEdit(title: string, callback: ClickHandler): UIElement {
+  return buttonSubtle("✎", title, callback);
+}
+/**
+ * Create a button that doesn't have the usual UI style for close and the like
```

Comment:
How is it different other than 'unusual'?

Reply:
They're for just icon buttons. I've renamed and added a comment.